Add YEAR_MONTH_INTERVAL and DURATION types (#496)
apache/parquet-format: LogicalTypes.md

    </tr>
</table>

> ### INTERVAL types

Author: updated

> #### INTERVAL

Reviewer: ```suggestion
#### INTERVAL

```

Reviewer 2: added new line

`INTERVAL` is *deprecated*. Please use `INTERVAL_YEAR_MONTH` and `INTERVAL_DAY_TIME`

> as a more precise representation per [ANSI SQL Standard](https://www.ibm.com/docs/en/informix-servers/14.10.0?topic=types-ansi-sql-standards-datetime-interval-values).

Reviewer: Is there a more official link rather than an IBM documentation?

Reviewer 2: I find an oracle definition, which might be better than the IBM one. updated

    </tr>
</table>

### Interval types

#### INTERVAL

> `INTERVAL` is *deprecated*. Please use `INTERVAL_YEAR_MONTH` and `INTERVAL_DAY_TIME`

Reviewer: From the discussion at https://docs.google.com/document/d/12ghQxWxyAhSQeZyy0IWiwJ02gTqFOgfYm8x851HZFLk, I think the existing Parquet interval types are not well-thought and not adopted widely (may be biased).

> `INTERVAL` is *deprecated*. Please use `INTERVAL_YEAR_MONTH` and `INTERVAL_DAY_TIME`

Reviewer 3: Ok, I read your comment there and I think you're right. The new type is less trivially convertible from Arrow's month_day_nano, though.

statistics should be saved for this type and if such non-compliant statistics
are found during reading, they must be ignored.

> #### INTERVAL_YEAR_MONTH

Reviewer 2: added new line

`INTERVAL_YEAR_MONTH` is used to represent a year-month time interval, such as
`4 years and 6 months`. It must annotate an `int32` that stores the total number
of months as a signed integer, which represents the interval and can be negative.
The time duration is independent of any timezone.

> #### INTERVAL_DAY_TIME

Reviewer: ```suggestion
#### INTERVAL_DAY_TIME

```

`INTERVAL_DAY_TIME` is used to represent a day-time time interval, such as

> `5 days, 10 hours and 30 minutes`. It must annotate and 16-byte `FIXED_LEN_BYTE_ARRAY`

Reviewer 4: We need a line that is like "As in ansi standard ... day_time [link or citiation) a day is always considered to be 24 hours (equivalent to X nanos)"

Author: i added one more paragraph at the end of this section. There is actually no documentation explicitly stated that 1 day is equal to 24 hours, those seems indicated from the value range definition of the SQL standard, so I linked the whole SQL standard doc here.

> `5 days, 10 hours and 30 minutes`. It must annotate and 16-byte `FIXED_LEN_BYTE_ARRAY`

Author: updated

statistics should be saved for this type and if such non-compliant statistics
are found during reading, they must be ignored.

#### INTERVAL_YEAR_MONTH

`INTERVAL_YEAR_MONTH` is used to represent a year-month time interval, such as
`4 years and 6 months`. It must annotate an `int32` that stores the total number
of months as a signed integer, which represents the interval and can be negative.
The time duration is independent of any timezone.

#### INTERVAL_DAY_TIME

`INTERVAL_DAY_TIME` is used to represent a day-time time interval, such as
`5 days, 10 hours and 30 minutes`. It must annotate a 16-byte `FIXED_LEN_BYTE_ARRAY`

> that stores the total number of nanoseconds representing the interval. The value is

Reviewer 3: Why call it "day-time" if it only stores nanoseconds? Please let's make the naming clear and non-misleading.

a signed integer and can be negative to indicate backward duration.
The time interval is independent of any timezone.

Based on the [ANSI SQL standard](https://docs.oracle.com/en/database/oracle/oracle-database/19/sqlrf/Data-Types.html#GUID-7690645A-0EE3-46CA-90DE-C96DF5A01F8F) definition of the INTERVAL data type and fields values, 

> an interval of 1 day is equivalent to 24 hours, regardless of the specific number of 

Reviewer 5: this detail should also probably be in parquet.thrift (reference to SQL standard probably isn't strictly necessary there, just how 1 day is defined).

Reviewer 6: This definition also turns this into what other systems often call a "duration" not an "interval" right? For instance this would map to pandas/datetime Timedelta and not pandas DateOffset.

Reviewer 5: @lidavidm you are correct.  SQL standard calls this type interval which is where the name comes from.

Author: i think the DayTimeInterval sometimes are referred as duration to be more clear on the unit used, and the pandas TimeDelta can be mapped to DayTimeInterval type, which is one of the biggest use case also. The YearMonth is clearly different as duration as @lidavidm pointed out.

@lidavidm 
```
this detail should also probably be in parquet.thrift (reference to SQL standard probably isn't strictly necessary there, just how 1 day is defined).
```
is your suggestion to remove the SQL reference here?

Reviewer 6: I didn't make that suggestion, to be clear ^_^ but I do think it would help to be clear (somewhere) about whether each type is a calendar interval/relative delta or a duration/absolute time delta

statistics should be saved for this type and if such non-compliant statistics
are found during reading, they must be ignored.

#### YEAR_MONTH_INTERVAL

`YEAR_MONTH_INTERVAL` is used to represent a year-month time interval, such as
`4 years and 6 months`. It must annotate an `int32` that stores the total number

> of months as a signed integer, which represents the interval and can be negative.

Reviewer 5: ```suggestion
of months as a signed integer, which represents the interval and can be negative. Some ANSI SQL systems only support intervals in the range of +/- 10,000 years, no constraint is placed on the exact value in Parquet.
```

Author: added description about the restriction.

> The time duration is independent of any timezone.

Author: sg! deleted

#### DURATION

`DURATION` is used to represent a span of time, such as `5 days`. It must 
annotate an `int64` value that stores the total number of time units for the 
duration. The value is a signed integer and can be negative to indicate backward duration.
The duration is purely a measure of time and is independent of any time zone.

The `DURATION` type takes `unit` as a parameter, and the value must be one of
`MILLIS`, `MICROS` or  `NANOS`.

> Duration of 1 day is defined exactly of 24 hours, regardless of the specific number of 

Reviewer 5: nit: I suggest moving this to be in the same a paragraph as `The duration is purely a measure of time and is independent of any time zone.` (split this paragraph off from the first paragraph).

Author: make sense. mentioned ANSI SQL standard in the description

statistics should be saved for this type and if such non-compliant statistics
are found during reading, they must be ignored.

#### YEAR_MONTH_INTERVAL

`YEAR_MONTH_INTERVAL` is used to represent a year-month time interval, such as
`4 years and 6 months`. It must annotate an `int32` that stores the total number
of months as a signed integer, which represents the interval and can be negative.

While ANSI SQL systems typically restrict supported intervals to a range of 
±10,000 years and enforce this constraint internally, the Parquet format 
does not impose any limitations on the interval values that may be stored.

#### DURATION

`DURATION` is used to represent a span of time, such as `5 days`. It must 
annotate an `int64` value that stores the total number of time units for the 

> duration. The value is a signed integer and can be negative to indicate backward 

Reviewer 4: Sounds good to me

statistics should be saved for this type and if such non-compliant statistics
are found during reading, they must be ignored.

#### YEAR_MONTH_INTERVAL

`YEAR_MONTH_INTERVAL` is used to represent a year-month time interval, such as
`4 years and 6 months`. It must annotate an `int32` that stores the total number
of months as a signed integer, which represents the interval and can be negative.

While ANSI SQL systems typically restrict supported intervals to a range of 
±10,000 years and enforce this constraint internally, the Parquet format 
does not impose any limitations on the interval values that may be stored.

#### DURATION

`DURATION` is used to represent a span of time, such as `5 days`. It must 
annotate an `int64` value that stores the total number of time units for the 
duration. The value is a signed integer, where a negative value indicates the
duration moves backward in time (e.g., -5 days means going backward for 5 days).  
The duration is purely a measure of time and is independent of any time zone.

The `DURATION` type takes `unit` as a parameter, and the value must be one of
`MILLIS`, `MICROS` or  `NANOS`.

> `Duration` can be used to represent DayTime Intervals as defined by ANSI SQL. In 

Reviewer: ```suggestion
`DURATION` can be used to represent day-time interval as defined by ANSI SQL. In 
```

Just to be consistent with above.
